### PR TITLE
Hydra

### PR DESCRIPTION
**Pull request by meyerzinn**
Building off of the work by @kminehart, I produced a more idiomatic and production-ready version of the Hydra chart. We agreed that the changes warranted a new PR.

Supersedes #1022

### CODE REVIEW

**Comment by kminehart** [comment]
Perfect, thank you for doing this.

**Comment by meyerzinn** [comment]
My pleasure. Thanks for getting the ball rolling on this @kminehart

**Comment by meyerzinn** [comment]
@Grillz if you'd like to review this one that would be amazing.

**Comment by unguiculus** [comment]
@k8s-bot ok to test

**Comment by meyerzinn** [comment]
Fixing.

File: incubator/hydra/Chart.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+- authentication
+- authorization
+- auth
+- oauth
+- jwt
+home: https://www.ory.am/products/hydra
+icon: https://github.com/ory/hydra/blob/master/docs/images/logo.png
+sources:
+- https://github.com/ory/hydra
+maintainers:
```
Comment (unguiculus) [inline incubator/hydra/Chart.yaml:18]: Use Github usernames. See e. g. https://github.com/kubernetes/charts/blob/master/stable/jenkins/Chart.yaml#L10-L13

File: incubator/hydra/templates/NOTES.txt [hunk 2]
```diff
@@ -0,0 +1,28 @@
+{{- if and (not .Values.postgresql.enabled) (not .Values.config.hydra.databaseURL ) }}
```
Comment (unguiculus) [inline incubator/hydra/templates/NOTES.txt]: Nit: Remove the space before the parenthesis (`.databaseURL )`).

File: incubator/hydra/values.yaml [hunk 3]
```diff
@@ -0,0 +1,103 @@
+image: "oryd/hydra"
+imageTag: "v0.8.6"
+imagePullPolicy: "Always"
```
Comment (unguiculus) [inline incubator/hydra/values.yaml]: Why `Always` as default?

File: incubator/hydra/templates/_helpers.tpl [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+Return the database URL (base64 encoded and quoted).
+If the postgresql dependency is enabled, it will return the URL for the postgres database.
+Otherwise, it will return the set value.
+*/}}
+{{- define "hydra.databaseURL" -}}
+{{- if .Values.postgresql.enabled -}}
+{{ $pgHost := printf "%s-%s" .Release.Name "postgresql" | trunc 63 | trimSuffix "-" }}
+{{- printf "postgres://%s:%s@%s/%s?sslmode=disable" .Values.postgresql.postgresUser .Values.postgresql.postgresPassword $pgHost .Values.postgresql.postgresDatabase | b64enc | quote -}}
+{{- else -}}
+{{- default "" .Values.config.hydra.database_url | b64enc | quote -}}
```
Comment (unguiculus) [inline incubator/hydra/templates/_helpers.tpl]: `database_url` -> `databaseURL`

You can remove the `default "" ` if you specify a default value in `values.yaml`.

**Review by unguiculus — CHANGES_REQUESTED** [review]
Thanks, this looks really good.

File: incubator/hydra/values.yaml (hunk 3 above)
Comment (meyerzinn) [reply incubator/hydra/values.yaml]: @unguiculus Hydra is currently pre-release (pre-v1, though generally stable) and pushes out frequent security or other patches. Thus, the intention here was to allow image tags to stabilize (i.e. a v1 tag that is v1.0-v1.X) while still pulling the Docker images, but it's not strictly required.

In the end, there's a reason this is configurable 😛.

File: incubator/hydra/templates/_helpers.tpl (hunk 4 above)
Comment (meyerzinn) [reply incubator/hydra/templates/_helpers.tpl]: @unguiculus It needs to be camel case to work as a source for environmental variables.

Comment (meyerzinn) [reply incubator/hydra/templates/_helpers.tpl]: Oh, I mistook the instance you were referring to. I'll change it.

File: incubator/hydra/Chart.yaml [hunk 5]
```diff
@@ -0,0 +1,20 @@
+appVersion: v0.8.6
+description: A security-first open source OAuth2 and OpenID Connect server for existing and new enterprise infrastructures.
+name: hydra
+version: 0.0.10
+keywords:
+- hdyra
```
Comment (pbarker) [inline incubator/hydra/Chart.yaml]: s/hdyra/hydra

File: incubator/hydra/Chart.yaml [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+- authentication
+- authorization
+- auth
+- oauth
+- jwt
```
Comment (pbarker) [inline incubator/hydra/Chart.yaml:11]: may as well add OIDC while were at it

**Comment by pbarker** [comment]
Aside from the semantic tag stuff, this looks really solid. Awesome work @20zinnm and @kminehart ! 👍

**Comment by meyerzinn** [comment]
@Grillz Added the tags. I believe that's a wrap 👍

File: incubator/hydra/Chart.yaml (hunk 6 above)
Comment (meyerzinn) [reply incubator/hydra/Chart.yaml:11]: Added with 9dfbaa8.

File: incubator/hydra/Chart.yaml (hunk 1 above)
Comment (meyerzinn) [reply incubator/hydra/Chart.yaml:18]: Fixed with 82c507c.

**Comment by meyerzinn** [comment]
@unguiculus I made the requested changes. Please let me know when you have a chance to look over it for final review.

File: incubator/hydra/values.yaml (hunk 3 above)
Comment (unguiculus) [reply incubator/hydra/values.yaml]: If I specify a version like `v0.9.1` I would not expect it to change. If this is actually the case, then `Always` should be a conscious choice. I'd suggest you change the default to `IfNotPresent`.

File: incubator/hydra/templates/NOTES.txt [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+{{- end }}
+
+To connect to Hydra, you can use:
+
+{{- if contains "NodePort" .Values.service.type }}
+  export NODE_IP=$(kubectl get nodes --namespace {{ .Release.Namespace }} -o jsonpath="{.items[0].status.addresses[0].address}")
+  echo http://$NODE_IP/
+{{- else if contains "LoadBalancer" .Values.service.type }}
+  export SERVICE_IP=$(kubectl get svc --namespace {{ .Release.Namespace }} {{ template "hydra.fullname" . }} -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
+  echo http://$SERVICE_IP/
```
Comment (unguiculus) [inline incubator/hydra/templates/NOTES.txt]: Port is missing

**Review by unguiculus — CHANGES_REQUESTED** [review]
I deployed it with defaults on a GKE cluster. Calling the load balancer URL on port 4444 returns this:

`{"error":{"code":500,"message":"Can not serve request over insecure http"}}`

Here's the log.

```
$ kubectl logs --namespace default $(kubectl get pods -l "app=hydra" -o jsonpath="{.items[0].metadata.name}") -c hydra                                                               ± ★ 〜 ⫘ (9dfbaa8)
Applying `ladon` SQL migrations...
Applied 0 `ladon` SQL migrations.
Applying `group` SQL migrations...
Applied 0 `group` SQL migrations.
Applying `client` SQL migrations...
Applied 0 `client` SQL migrations.
Applying `oauth2` SQL migrations...
Applied 0 `oauth2` SQL migrations.
Applying `jwk` SQL migrations...
Applied 0 `jwk` SQL migrations.
Migration successful! Applied a total of 0 SQL migrations.
Migration successful!
time="2017-06-13T16:40:40Z" level=info msg="Connecting with postgres://*:*@ironic-stoat-postgresql/hydra?sslmode=disable"
time="2017-06-13T16:40:40Z" level=info msg="Connected to SQL!"
time="2017-06-13T16:40:40Z" level=info msg="Setting up telemetry - for more information please visit https://ory.gitbooks.io/hydra/content/telemetry.html"
time="2017-06-13T16:40:40Z" level=info msg="Setting up http server on :4444"
time="2017-06-13T16:40:40Z" level=info msg="TLS termination enabled, disabling https."
time="2017-06-13T16:40:44Z" level=info msg="started handling request" method=GET remote="10.28.2.1:37524" request="/health"
time="2017-06-13T16:40:44Z" level=warning msg="Could not serve http connection" error="X-Forwarded-Proto header is missing"
time="2017-06-13T16:40:44Z" level=error msg="An error occurred while handling a request" code=502 details=[] error="Can not serve request over insecure http" reason= request-id= status= trace="Stack trace:
github.com/ory/hydra/cmd/server.(*Handler).rejectInsecureRequests
	/go/src/github.com/ory/hydra/cmd/server/handler.go:165
github.com/ory/hydra/cmd/server.(*Handler).(github.com/ory/hydra/cmd/server.rejectInsecureRequests)-fm
	/go/src/github.com/ory/hydra/cmd/server/handler.go:69
github.com/ory/hydra/vendor/github.com/urfave/negroni.HandlerFunc.ServeHTTP
	/go/src/github.com/ory/hydra/vendor/github.com/urfave/negroni/negroni.go:24
github.com/ory/hydra/vendor/github.com/urfave/negroni.middleware.ServeHTTP
	/go/src/github.com/ory/hydra/vendor/github.com/urfave/negroni/negroni.go:33
github.com/ory/hydra/vendor/github.com/urfave/negroni.(middleware).ServeHTTP-fm
	/go/src/github.com/ory/hydra/vendor/github.com/urfave/negroni/negroni.go:33
github.com/ory/hydra/vendor/github.com/meatballhat/negroni-logrus.(*Middleware).ServeHTTP
	/go/src/github.com/ory/hydra/vendor/github.com/meatballhat/negroni-logrus/middleware.go:135
github.com/ory/hydra/vendor/github.com/urfave/negroni.middleware.ServeHTTP
	/go/src/github.com/ory/hydra/vendor/github.com/urfave/negroni/negroni.go:33
github.com/ory/hydra/vendor/github.com/urfave/negroni.(middleware).ServeHTTP-fm
	/go/src/github.com/ory/hydra/vendor/github.com/urfave/negroni/negroni.go:33
github.com/ory/hydra/metrics.(*MetricsManager).ServeHTTP
	/go/src/github.com/ory/hydra/metrics/middleware.go:133
github.com/ory/hydra/vendor/github.com/urfave/negroni.middleware.ServeHTTP
	/go/src/github.com/ory/hydra/vendor/github.com/urfave/negroni/negroni.go:33
github.com/ory/hydra/vendor/github.com/urfave/negroni.(*Negroni).ServeHTTP
	/go/src/github.com/ory/hydra/vendor/github.com/urfave/negroni/negroni.go:73
github.com/ory/hydra/vendor/github.com/gorilla/context.ClearHandler.func1
	/go/src/github.com/ory/hydra/vendor/github.com/gorilla/context/context.go:141
net/http.HandlerFunc.ServeHTTP
	/usr/local/go/src/net/http/server.go:1942
net/http.serverHandler.ServeHTTP
	/usr/local/go/src/net/http/server.go:2568
net/http.(*conn).serve
	/usr/local/go/src/net/http/server.go:1825
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:2197" writer=JSON
time="2017-06-13T16:40:44Z" level=info msg="completed handling request" measure#hydra.example.com.latency=393805 method=GET remote="10.28.2.1:37524" request="/health" status=502 text_status="Bad Gateway" took=393.805µs
```

File: incubator/hydra/templates/NOTES.txt (hunk 7 above)
Comment (meyerzinn) [reply incubator/hydra/templates/NOTES.txt]: @unguiculus I think the idea here is that it's proxied to :80, since it's a load balancer, no?

**Comment by meyerzinn** [comment]
@unguiculus When behind a load balancer, it needs a correct HTTPS_ALLOW_TERMINATION_FROM environmental variable. However, I've found that for some reason GCE load balancers don't send the correct headers downstream, so you can manually disable HTTPS behind a load balancer by adding `--dangerous-force-http`.

I wasn't able to identify the issue but I think it has more to do with GCE than Hydra.

**Comment by meyerzinn** [comment]
Oh, @unguiculus Hydra cannot work with ingress out of the box because it requires SSL or SSL termination. Try setting `config.http.allowTerminationFrom` to whatever range GCE uses (I believe it's `10.0.0.0/8`.

I'll look into it more, but generally speaking Hydra shouldn't be publically accessible without SSL.

**Comment by meyerzinn** [comment]
So what I've found out is that GCE load balancer for whatever reason are not adding the correct proxy headers. Thus, try NodePort and disable SSL termination.

**Comment by meyerzinn** [comment]
Or use an ingress and properly configure it.

File: incubator/hydra/templates/NOTES.txt (hunk 7 above)
Comment (unguiculus) [reply incubator/hydra/templates/NOTES.txt]: `service.port`is `4444`. That's what is used. This port must be added here.

**Comment by unguiculus** [comment]
We have the general assumption that a chart should come up successfully with defaults. Is this somehow feasible? You cannot expect us to do more extensive tests.

**Comment by meyerzinn** [comment]
@unguiculus I don't think that's feasible for a secure-by-default Hydra installation. Due to the nature of Hydra (which necessitates SSL/TLS for basic functionality) I don't think this chart should work with defaults. Of course, it could be configured to dump SSL/TLS (with the `--dangerous-force-http` flag), but that makes it more difficult to use it securely.

I could add in the notes that Hydra can only be exposed publicly (via the Ingress) with SSL, and that for internal use ClusterIP with `--dangerous-force-http` could work. Or should I add a boolean value?

**Comment by meyerzinn** [comment]
I'm going on a trip for about 12 days and won't be in internet range. Let's pick this up when I get back, unless someone else wants to make the changes.

**Comment by unguiculus** [comment]
Marking as stale. Please update within the next week.

**Comment by meyerzinn** [comment]
Sorry, been really busy. Will see if I find some time soon.

**Comment by meyerzinn** [comment]
@unguiculus Not sure what the conclusion was as to whether Hydra should be secure-by-default. It's not possible to set up Hydra publicly to be secure because of the transport security requirement, but the ingress is not the default exposure (I think? been a month, will need to check) so it should come up fine by default.

If the ingress is the default, just taking that off the default and leaving it as a service solves the issue. However, that means Hydra is unable to perform basic functionality (serve HTTP redirects to control the consent flow).

File: incubator/hydra/README.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
+# Hydra
+[Hydra](https://github/ory/hdya) offers OAuth 2.0 and OpenID Connect Core 1.0 capabilities as a service. Hydra is different, because it works with any existing authentication infrastructure, not just LDAP or SAML. By implementing a consent app (works with any programming language) you build a bridge between Hydra and your authentication infrastructure.
```
Comment (unguiculus) [inline incubator/hydra/README.md:2]: The Github link is incorrect.

**Review by unguiculus — CHANGES_REQUESTED** [review]
OK, I'd suggest you add some more documentation to the README and also some prominent message to the NOTES. Here's an example for NOTES from Datadog to illustrate what I mean.

https://github.com/kubernetes/charts/blob/master/stable/datadog/templates/NOTES.txt

We should just make sure we fail gracefully and the user gets directions how they can make things work.

**Comment by unguiculus** [comment]
Marking this as stale. Please update within one week.

**Comment by unguiculus** [comment]
Closing this as stale. Feel free to reopen if/when you decide to work on this again.

**Comment by alonl** [comment]
@20zinnm I'll be very happy to see this on air :)

**Comment by haf** [comment]
@unguiculus How about merging this and letting the community work on it? The author is right that Google's L7 LB has issues with SNI and requires manual intervention to bootstrap TLS certificates; and at the same time, a security-critical piece of infrastructure like Ory Hydra has a functional requirement not to be accessible unless the channel is encrypted. I would deploy this with certmanager and let's encrypt via nginx-ingress, traefik or envoy ingresses; and that's out of scope for this chart.

**Comment by aeneasr** [comment]
In case this was missed, we have now official helm charts: https://k8s.ory.sh/helm/

**Pull request closed without merging**